hazelcast/src/main/java/com/hazelcast/collection/impl/collection/CollectionService.java

```diff
@@ -88,9 +91,25 @@ public void shutdown(boolean terminate) {
 
     @Override
     public void destroyDistributedObject(String name) {
-        CollectionContainer container = getContainerMap().remove(name);
+        final CollectionContainer container = getContainerMap().remove(name);
+        final int partitionId = nodeEngine.getPartitionService().getPartitionId(StringPartitioningStrategy.getPartitionKey(name));
         if (container != null) {
-            container.destroy();
+            if (Thread.currentThread().getClass() == PartitionOperationThread.class) {
```

Comment:
seems `partitionId` check is missing here. 
`container.destroy()` should run inside `PartitionOperationThread` of matching `partitionId`.

Reply:
If this is already running in PartitionOperationThread, doesn't it mean it runs in the correct partition? If somebody delegated this to a partition thread, I assume the caller must have called it with correct partition id.

Reply:
nothing forces it to run on the correct partition id, better to check.